### PR TITLE
fix #60: Creating pipeline insertion module

### LINKED ISSUE #60: Create  the pipeline insertion module
## Issue Description

To help the data scientist to build and push pipelines and models to the blob storage a script was drafted to create this functionality.
However, it is not complete and does not reflect code standards. A complete refactoring is needed to give a more complete tool to help our data scientists.

### Work to do

Create a command line application to upload the JSON file containing the pipeline and the models to the Blob Storage. Version one should be simple and ask the user the path of the document to insert in the blob storage.

### Acceptance Criteria

- The pipeline insertion module asks for the path of the JSON file
- It correctly crypts sensible data before uploading it to the blob storage
- Documentation is clear and complete on how to use the module and insert pipelines into the blob storage

### Tasks

- [x] Create a command line interface script that asks for the JSON path
- [x] Validate that the file is a JSON and contains models and pipelines
- [x] Create a function to crypt sensible data. *That encrypting needs to be decrypted by the backend when retrieving the pipelines and models information*
- [x] Create a pipeline insertion function that notifies the users of any errors or success
- [x] document the usage of the module
- [x] Check that a default pipeline is set by the user
  - https://github.com/ai-cfia/nachet-frontend/issues/111

### PR DESCRIPTION
**Pull request by MaxenceGui**
- [x] Create a command line interface script that asks for the file path
- [x] Validate the structure of the file and its extension
- [x] Crypt sensible data. 
- [x] Notify the user when errors happen or if the operation succeed
- [x] Document the usage of the module
- [x] Implement tests in `tests` folder

### CODE REVIEW

**Comment by MaxenceGui** [comment]
This work is based on branch 51-implementing-2-models from
* #55

Change starts at commit `fixes #60`

@amirardalan9473 if you can review the `pipeline_template.yaml` and validate that all the data here are relevant I would appreciate it thank you

**Review by SonOfLope — APPROVED** [review]
Approving changes made to .github folder.

File: TESTING.md
Comment (RussellJimmies) [inline TESTING.md]: quart

File: TESTING.md
Comment (RussellJimmies) [inline TESTING.md]: quart

File: app.py
Comment (RussellJimmies) [inline app.py]: Should update the docstring if the function no longer does the caching

File: docs/nachet-inference-documentation.md
Comment (RussellJimmies) [inline docs/nachet-inference-documentation.md]: pipeline

File: docs/nachet-inference-documentation.md
Comment (RussellJimmies) [inline docs/nachet-inference-documentation.md]: retrieve**s** the data [...] and store**s** the pipeline

File: docs/nachet-inference-documentation.md
Comment (RussellJimmies) [inline docs/nachet-inference-documentation.md]: This **is** the variable that feed**s**

File: docs/nachet-model-documentation.md
Comment (RussellJimmies) [inline docs/nachet-model-documentation.md]: Usually in AI, the top 5 result**s** are use**d** to measure the accuracy of a model. If the correct result is **in** the top 5, then the prediction is considered true.

File: docs/nachet-model-documentation.md
Comment (RussellJimmies) [inline docs/nachet-model-documentation.md]: What do you think of this? 

This is useful in **the** case w**h**ere the user might have a need for multiple guesses for one inference.

**Review by RussellJimmies — CHANGES_REQUESTED** [review]
Mainly reviewed grammar

File: app.py
Comment (MaxenceGui) [reply app.py]: Done

File: TESTING.md
Comment (MaxenceGui) [reply TESTING.md]: done

File: docs/nachet-inference-documentation.md
Comment (MaxenceGui) [reply docs/nachet-inference-documentation.md]: Done

File: docs/nachet-inference-documentation.md
Comment (MaxenceGui) [reply docs/nachet-inference-documentation.md]: Done

File: docs/nachet-inference-documentation.md
Comment (MaxenceGui) [reply docs/nachet-inference-documentation.md]: Done

File: docs/nachet-model-documentation.md
Comment (MaxenceGui) [reply docs/nachet-model-documentation.md]: Done

File: docs/nachet-model-documentation.md
Comment (MaxenceGui) [reply docs/nachet-model-documentation.md]: Like it. Done

File: .devcontainer/devcontainer.json
Comment (elwoujdi) [inline .devcontainer/devcontainer.json]: In a discussion I had with @rngadam, he stated that he prefers not to add `act` on backend repositories and instead leave that to the devops repositories.

File: README.md
Comment (elwoujdi) [inline README.md]: For consistency, add more descriptive details.

File: .devcontainer/devcontainer.json
Comment (MaxenceGui) [reply .devcontainer/devcontainer.json]: Wich one do I need to not have? All of the features? Or just a specific one?

File: README.md
Comment (MaxenceGui) [reply README.md]: Done

File: .devcontainer/devcontainer.json
Comment (elwoujdi) [reply .devcontainer/devcontainer.json]: If you use docker in your back-end repo you can keep `"ghcr.io/devcontainers/features/docker-in-docker:2"` for Codespaces. Otherwise, delete the 2 of them.

Comment (MaxenceGui) [reply .devcontainer/devcontainer.json]: Done

**Review by RussellJimmies — APPROVED** [review]

**Review by rngadam — APPROVED** [review]

File: docs/nachet-inference-documentation.md
Comment (rngadam) [inline docs/nachet-inference-documentation.md]: pass the cipher_suite and the result_json as the caller. don't rely on global environment variables being retrieved within your lower level functions.

"mock" should not be referenced within functions, that's a test facility

**Review by rngadam — CHANGES_REQUESTED** [review]

File: docs/nachet-inference-documentation.md
Comment (MaxenceGui) [reply docs/nachet-inference-documentation.md]: I retrieve the mock here on PR55:

https://github.com/ai-cfia/nachet-backend/blob/51-implementing-2-models/docs/nachet-inference-documentation.md#in-the-code

Once PR55 is merge, I will correct it here with lower function receiving the environment variables

Comment (MaxenceGui) [reply docs/nachet-inference-documentation.md]: Done

image

image

**Comment by MaxenceGui** [comment]
@amirardalan9473 If you can do a quick review
The important file for you are pipeline_template.yaml and the pipelines_version_insertion.py and of course the documentation

File: app.py
Comment (amirardalan9473) [inline app.py:93]: why these dimensions?

File: app.py
Comment (amirardalan9473) [inline app.py:163]: what does this json contain? "seeds/all.json"

File: app.py
Comment (MaxenceGui) [reply app.py:93]: It's the standard dimension for an HD image and will act as default on Nachet if  the `NACHET_VALID_DIMENSION` environment variable is not set or is not able to load.

We can always modify these default value if we want bigger or smaller image

File: app.py
Comment (MaxenceGui) [reply app.py:163]: Here is the link [nachet-data/seed/all.,json](https://github.com/ai-cfia/nachet-data/blob/main/seeds/all.json),  it is data from all the seeds collected by Jolan.

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: models vs model_name? is model_name supposed to be a list of all the models?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: if were querying from DB after francoi is done, it wont necessarily have a name. Perhaps the query? or a description? in addition

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: is this necessary?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: what does this mean exactly?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: ?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: accuracy?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: ?

File: pipeline_template.yaml
Comment (amirardalan9473) [inline pipeline_template.yaml]: same as the previous dataset comment. also whats the difference?

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: Models is a list containing the pipeline model and  `model_name ` is to fit the frontend until they use `pipeline_name` , then we can delete the attribute

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: Yes, it is necessary to build the `headers` for the HTTP requests.

https://github.com/ai-cfia/nachet-backend/blob/0b4e0a4a018040d012fd321ab01ae44e2197927d/model/seed_detector.py#L79

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: We could go with `dataset_description` to give an idea of what the dataset is.

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: `content_type` is a string representation of what is sent to the model, in our case its always an `application/json`, so it's basically a way for the model to say to the backend what he needs to perform the inference.

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: It's an artefact and I will delete it, thank you for pointing it out!

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: Certainly

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: This is the dataset for the model, and previously it was for the pipeline containing the models. It might be a bit repetitive and unnecessary for pipeline,

File: pipeline_template.yaml
Comment (MaxenceGui) [reply pipeline_template.yaml]: Took it from here:

["identifiable": []](https://github.com/ai-cfia/nachet-model/blob/9443406e59575630cc90445aa8dc979efc40634c/model_endpoints_metadata.json#L12)

In nachet-model, if you don't what it is, then I guess we don't need this attribute

**Review by amirardalan9473 — APPROVED** [review]

**Comment by MaxenceGui** [comment]
@CFIALeronB and @ChromaticPanic

After the merge of this PR, next version of pipeline will have a `default`, and will have the attribute `pipeline_name` instead of `model_name` so just be careful to do the modification necessary in the frontend.

* https://github.com/ai-cfia/nachet-frontend/issues/111

**Pull request merged**
